lib/fizzy/parser_expr.cpp

```diff
@@ -162,21 +179,32 @@ parser_result<Code> parse_expr(
         auto& frame = control_stack.top();
         const auto& metrics = metrics_table[opcode];
 
-        if (!frame.unreachable)
+        if ((frame.stack_height - frame.parent_stack_height) < metrics.stack_height_required)
         {
-            if ((frame.stack_height - frame.parent_stack_height) < metrics.stack_height_required)
+            // Stack is polymorphic after unreachable instruction: underflow is ignored,
+            // but we need to count stack growth to detect extra values at the end of the block.
+            if (frame.unreachable)
+            {
+                // Add instruction's output values only.
+                frame.stack_height = frame.parent_stack_height + metrics.stack_height_required +
+                                     metrics.stack_height_change;
+            }
+            else
                 throw validation_error{"stack underflow"};
-
+        }
+        else
+        {
             // Update code's max_stack_height using frame.stack_height of the previous instruction.
             // At this point frame.stack_height includes additional changes to the stack height
             // if the previous instruction is a call/call_indirect.
             // This way the update is skipped for end/else instructions (because their frame is
             // already popped/reset), but it does not matter, as these instructions do not modify
             // stack height anyway.
-            code.max_stack_height = std::max(code.max_stack_height, frame.stack_height);
-        }
+            if (!frame.unreachable)
```

Comment:
If we pull this `max_stack_height` out of outer `if`, it looks like we can unify stack height update/checks for instructions and for calls, not sure if it's worth it. The error message for stack underflow would not be specific for calls then...

Reply:
I think this can be done in separate PR to confirm coverage of this version is full.